src/compiler/checker.ts

```diff
@@ -2663,7 +2746,170 @@ namespace ts {
+        */
+        function resolveDefineModule(symbol: Symbol): Type {
+            // Can't meaningfully define the same module more than once
+            if (symbol.declarations.length !== 1) {
+                return unknownType;
+            }
+
+            // If the invocation of 'define' has zero args ('define()')... we
+            // shouldn't really be here, but it's meaninguless
+            let callExpr = <CallExpression>symbol.declarations[0];
+            if (callExpr.arguments.length === 0) {
+                Debug.fail("Should not have a zero-arg define call");
+                return unknownType;
+            }
+
+            // If the last arg isn't a function expr, just resolve
+            // its type and define that as the shape of the module
+            let lastArg = <FunctionExpression>callExpr.arguments[callExpr.arguments.length - 1];
+            if (!isFunctionLike(lastArg)) {
+                let resultType = getTypeOfExpression(lastArg);
+                // If this type is a function type, we want to use its return type since
+                // it's going to get invoked anyway
+                let signatures = getSignaturesOfType(resultType, SignatureKind.Call);
+                if (signatures.length > 0) {
+                    return getReturnTypeOfSignature(signatures[0]);
+                }
+                else {
+                    return resultType;
+                }
+            }
+
+            // If there are any expressionful return statements in this function,
+            // we'll use those as the shape of the module
+            let bodyReturnType = getReturnTypeFromBody(lastArg);
+            if (bodyReturnType !== voidType) {
+                return bodyReturnType;
+            }
+
+            // Collect assignments in the body to 'module.exports' or 'exports.propName' or 'module.exports.propName'
+            let assignedModuleType: Type = undefined;
+            let exportAssignedProperties: Symbol[] = [];
+            let exportAssignedPropTable: SymbolTable = {};
+
+            if (cjsModuleType === undefined || cjsExportsType === undefined) {
+                return unknownType;
+            }
+
+            traverse(lastArg.body);
+
+            // Someone assigned to 'module.exports', so ignore other assignments
+            if (assignedModuleType) {
+                return assignedModuleType;
+            }
+
+            // Otherwise return the collected property assignments
+            if (exportAssignedProperties.length > 0) {
+                return createAnonymousType(lastArg.symbol,
+                        exportAssignedPropTable,
+                        /*callSignatures*/ [],
+                        /*constructSignatures*/ [],
+                        /*stringIndexType*/ undefined,
+                        /*numberIndexType*/ undefined);
+            }
+
+            // Empty module?
+            return emptyObjectType;
+
+            function traverse(node: Node) {

```

Comment:
This inference is based on the type checker determining which expression(s) represent `module` (`if (operandType === cjsModuleType`), so we don't yet know during the bind phase whether a given property assignment represents an exported member.

Is there still some meaningful reuse of symbols that can be done given that constraint?